frontend/src/app/location-edit/location-edit.component.ts

```diff
@@ -0,0 +1,150 @@
+ */
+enum CrudOperation {
+    Update = 'Update',
+    Create = 'Create'
+}
+
+@Component({
+    selector: 'app-location-edit',
+    templateUrl: './location-edit.component.html',
+    styleUrls: ['./location-edit.component.css', '../card.css']
+})
+/**
+ * A Component for Creating and Updating singular locations.
+ */
+export class LocationEditComponent {
+    nameFormControl: FormControl = REQUIRED_NAME;
+    nameErrorMessage = REQUIRED_NAME_ERROR_MESSAGE;
+    typeFormControl = new FormControl('', [Validators.required]);
+    crudOperation: CrudOperation;
+    initialLocation: Location;
+    childrenLocations: Location[];
+    id: number;
+    config: Config;
+    @ViewChild('parentLocation') parentLocation: SelectLocationComponent;
+
+
+    constructor(private route: ActivatedRoute,
+                private configService: ConfigService,
+                private http: HttpClient,
+                private router: Router) {
+        this.config = this.configService.getConfig();
+        this.route.paramMap.subscribe(
+            params => {
+                this.id = +(params.get('id'));
+
+                const locationResolver = this.route.snapshot.data.locationResolver;
+                this.initialLocation = locationResolver._embedded.locations.find(location => location.id === this.id);
+                this.crudOperation = this.initialLocation ? CrudOperation.Update : CrudOperation.Create;
+                this.childrenLocations = [];
+
+                if (this.crudOperation === CrudOperation.Create) {
+                    this.initialLocation = {
+                        address: '',
+                        children: null,
+                        country: '',
+                        id: null,
+                        name: '',
+                        parent: null,
+                        type: '',
+                        zip: '',
+                    };
+                } else if (this.crudOperation === CrudOperation.Update) {
+                    this.http.get(this.config.getBackendUrl() + '/locations/' + this.id + '/parent')
+                        .subscribe((res: Response) => {
+                            this.parentLocation.setSelectedLocation(res);
+                        });
+
+                    this.http.get(this.config.getBackendUrl() + '/locations/' + this.id + '/children')
+                        .subscribe(res => {
+                            this.childrenLocations = sortArray(getPayload(res).locations, 'name');
+                        });
+                }
+
+                this.nameFormControl.setValue(this.initialLocation.name);
+                this.typeFormControl.setValue(this.initialLocation.type);
+            }
+        );
+    }
+
+
+    typeErrorMessage() {
+        return this.typeFormControl.hasError('required') ? 'You must enter a value' : '';
+    }
+
+
+    submit(name: string, address: string, zip: string, country: string, type: string) {
+        if (this.nameFormControl.invalid) {
```

Comment:
Nice assertions! These will never be triggered though, because the button will not be available until required fields are filled though.

Reply:
Those are mostly defensive checks. It prevents the really stupid malicious users who see a disabled button, open up chromes editor, and then enable it just to see what the fuck happens if they submit data anyway.

But I'm sure more intelligent malicious users could get around that easily, since it's still on the frontend. Can remove if you want.